samza-azure/src/main/java/org/apache/samza/AzureJobCoordinator.java

```diff
@@ -0,0 +1,402 @@
+        LOG.info("Leader detected barrier completion.");
+        state = BarrierState.END.name() + " " + nextJMVersion;
+      }
+      leaderBlob.publishBarrierState(state, azureLeaderElector.getLeaseId().get());
+      leaderBarrierSF.cancel(true);
+    };
+  }
+
+  /**
+   * Creates a listener for LivenessCheckScheduler class.
+   * Invoked by the leader when the list of active processors in the system changes.
+   * @return an instance of SchedulerStateChangeListener.
+   */
+  private SchedulerStateChangeListener createLivenessListener(AtomicReference<List<String>> liveProcessors) {
+    return () -> {
+      LOG.info("Leader detected change in list of live processors.");
+      doOnProcessorChange(liveProcessors.get());
+    };
+  }
+
+  /**
+   * Creates a listener for JMVersionUpgradeScheduler class.
+   * Invoked when the processor detects a job model version upgrade on the blob.
+   * Stops listening for job model version upgrades until rebalancing achieved.
+   * @return an instance of SchedulerStateChangeListener.
+   */
+  private SchedulerStateChangeListener createJMVersionUpgradeListener() {
+    return () -> {
+      LOG.info("Job model version upgrade detected.");
+      versionUpgradeDetected.getAndSet(true);
+      onNewJobModelAvailable(leaderBlob.getJobModelVersion());
+    };
+  }
+
+  /**
+   * Creates a listener for LeaderLivenessCheckScheduler class.
+   * Invoked when an existing leader dies. Enables the JC to participate in leader election again.
+   * @return an instance of SchedulerStateChangeListener.
+   */
+  private SchedulerStateChangeListener createLeaderLivenessListener() {
+    return () -> {
+      LOG.info("Existing leader died.");
+      azureLeaderElector.tryBecomeLeader();
+    };
+  }
+
+
+  private JobModel generateNewJobModel(List<String> processors) {
+    return JobModelManager.readJobModel(this.config, Collections.emptyMap(), null, streamMetadataCache,
+        processors);
+  }
+
+  /**
+   * Called only by the leader, either when the processor becomes the leader, or when the list of live processors changes.
+   * @param currentProcessorIds New updated list of processor IDs which caused the rebalancing.
+   */
+  private void doOnProcessorChange(List<String> currentProcessorIds) {
+    // if list of processors is empty - it means we are called from 'onBecomeLeader'
+
+    String nextJMVersion;
+    String prevJMVersion = currentJMVersion.get();
+    JobModel prevJobModel = jobModel;
+    AtomicBoolean barrierTimeout = new AtomicBoolean(false);
+
+    nextJMVersion = Integer.toString(Integer.valueOf(prevJMVersion) + 1);
+    if (currentProcessorIds.isEmpty()) {
+      currentProcessorIds = new ArrayList<>(table.getActiveProcessorsList(currentJMVersion));
+    } else {
+      //Check if previous barrier not reached, then previous barrier times out.
+      String blobJMV = leaderBlob.getJobModelVersion();
+      if (blobJMV != null && Integer.valueOf(blobJMV) > Integer.valueOf(prevJMVersion)) {
```

Comment:
It is to be noted that the local variable currentJMVersion is only set when rebalancing completes successfully in onNewJobModelConfirmed(). The expected initial value of prevJMVersion/currentJMVersion when entering this function should be the same as the blobJMV. This is because this function is only invoked by the leader, and it is the leader that will introduce the difference between these two values by publishing the new blobJMV > prevJMVersion.